firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,174 @@
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Optional;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ *
+ * <p>IMPORTANT: each recorder holds a reference to an Activity, so it is very important to
+ * dereference each recorder at or before its Activity's onDestroy. Similar for Fragments.
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+
+  private final Activity activity;
+  private final FrameMetricsAggregator frameMetricsAggregator;
+  private final Map<Fragment, PerfFrameMetrics> subTraceMap;
+
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity.
+   *
+   * @param activity the activity that the recorder is collecting data from.
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new HashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      Map<Fragment, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.frameMetricsAggregator = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window. */
+  public void start() {
+    if (isRecording) {
+      logger.warn(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    frameMetricsAggregator.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window.
+   *
+   * @return FrameMetrics accumulated during the current recording.
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.warn("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    if (!subTraceMap.isEmpty()) {
+      logger.warn(
+          "Sub-traces are still ongoing! Sub-traces should be stopped first before stopping Activity screen trace.");
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      frameMetricsAggregator.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug(
+          "View not hardware accelerated. Unable to collect FrameMetrics. %s", err.toString());
+      return Optional.absent();
+    }
+    frameMetricsAggregator.reset();
+    isRecording = false;
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording.
+   *
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Fragment key) {
```

Comment:
Can we rename this to `startFragment(Fragment fragment)`. Since the signature is locked, it makes sense to use the same terminology. We can generalize this later. Over engineering is equally challenging to maintain.